crates/aiken-lang/src/tipo/environment.rs

```diff
@@ -1068,11 +1071,73 @@ impl<'a> Environment<'a> {
                     tipo,
                 );
 
-                if !public && (kind.is_lib() || !VALIDATOR_NAMES.contains(&name.as_str())) {
+                if !public && kind.is_lib() {
                     self.init_usage(name.clone(), EntityKind::PrivateFunction, *location);
                 }
             }
 
+            Definition::Validator(Validator {
+                fun,
+                location,
+                params,
+                ..
+            }) if kind.is_validator() => {
+                assert_unique_value_name(names, &fun.name, location)?;
+
+                // Create the field map so we can reorder labels for usage of this function
+                let mut field_map = FieldMap::new(fun.arguments.len() + params.len(), true);
+
+                // Chain together extra params and function.arguments
+                for (i, arg) in params.iter().chain(fun.arguments.iter()).enumerate() {
+                    field_map.insert(arg.arg_name.get_label().clone(), i, &arg.location)?;
+                }
+
+                let field_map = field_map.into_option();
+
+                // Construct type from annotations
+                let mut hydrator = Hydrator::new();
+
+                hydrator.permit_holes(false);
+
+                let mut arg_types = Vec::new();
+
+                for arg in params.iter().chain(fun.arguments.iter()) {
+                    let tipo = hydrator.type_from_option_annotation(&arg.annotation, self)?;
+
+                    arg_types.push(tipo);
+                }
+
+                let return_type =
+                    hydrator.type_from_option_annotation(&fun.return_annotation, self)?;
+
+                let tipo = function(arg_types, return_type);
+
+                // Keep track of which types we create from annotations so we can know
+                // which generic types not to instantiate later when performing
+                // inference of the function body.
+                hydrators.insert(fun.name.clone(), hydrator);
+
+                // Insert the function into the environment
+                self.insert_variable(
+                    fun.name.clone(),
+                    ValueConstructorVariant::ModuleFn {
```

Comment:
I wonder if that is useful / desired. Since we never want to allow _calling_ validators from another function, or to export them from a module; we may not even register them in the type-checker environment in the end 🤔. This might perhaps generate slightly confusing errors if one tries to call a validator, but it also removes another class of errors / checks we need to perform to make sure one doesn't call a validator as a function.

Reply:
I thought this too, but then realized a use case @alessandrokonrad has in nebula where he calls his validators in unit tests, it would still be nice to allow people to call these functions in that case.

Reply:
I'm not aware, but good idea actually :D

Reply:
See what you've done @rvcas.